feat: added C implementaion for `@stdlib/stats/base/dists/rayleigh/kurtosis`

DhruvArvindSingh opened the pull request:
Resolves #3800 

## Description

> What is the purpose of this pull request?

This pull request:

-   #3800 

## Related Issues

> #3800 

This pull request:

-   resolves #3800 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

DhruvArvindSingh committed:
feat: added C implementaion for @stdlib/stats/base/dists/rayleigh/kurtosis

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Planeshifter commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/kurtosis/README.md:
```diff
@@ -116,6 +116,88 @@ for ( i = 0; i < 10; i++ ) {
+
+#### stdlib_base_dists_rayleigh_kurtosis( sigma )
+
+Evaluates the skewness for a Rayleigh distribution.
```
```suggestion
Returns the [excess kurtosis][kurtosis] of a [Rayleigh][rayleigh-distribution] distribution with scale parameter `sigma`.
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/kurtosis/lib/native.js:
```diff
@@ -0,0 +1,61 @@
+
+/**
+* Returns the excess kurtosis of a Rayleigh distribution.
+*
+* @param {NonNegativeNumber} sigma - scale parameter
```
```suggestion
*
* @private
* @param {NonNegativeNumber} sigma - scale parameter
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/kurtosis/include/stdlib/stats/base/dists/rayleigh/kurtosis.h:
```diff
@@ -0,0 +1,38 @@
+#endif
+
+/**
+* Returns the kurtosis for a Rayleigh distribution.
```
```suggestion
* Returns the excess kurtosis of a Rayleigh distribution.
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/kurtosis/examples/c/example.c:
```diff
@@ -0,0 +1,38 @@
+	int i;
+
+	for ( i = 0; i < 25; i++ ) {
+		sigma = random_uniform( 0 , 10.0 );
```
```suggestion
		sigma = random_uniform( 0.0, 10.0 );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/kurtosis/README.md:
````diff
@@ -116,6 +116,88 @@ for ( i = 0; i < 10; i++ ) {
+        printf( "σ: %lf, f(x;σ): %lf\n", sigma, y );
+    }
+}
+```
````
````suggestion
</section>

<!-- /.usage -->

<!-- C API usage notes. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="notes">

</section>

<!-- /.notes -->

<!-- C API usage examples. -->

<section class="examples">

### Examples

```c
#include "stdlib/stats/base/dists/rayleigh/kurtosis.h"
#include <stdlib.h>
#include <stdio.h>

static double random_uniform( const double min, const double max ) {
    double v = (double)rand() / ( (double)RAND_MAX + 1.0 );
    return min + ( v*(max-min) );
}

int main( void ) {
    double sigma;
    double y;
    int i;

    for ( i = 0; i < 25; i++ ) {
        sigma = random_uniform( 0.0, 10.0 );
        y = stdlib_base_dists_rayleigh_skewness( sigma );
        printf( "σ: %lf, Kurt(X,σ): %lf\n", sigma, y );
    }
}
```
````

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/kurtosis/examples/c/example.c:
```diff
@@ -0,0 +1,38 @@
+	for ( i = 0; i < 25; i++ ) {
+		sigma = random_uniform( 0 , 10.0 );
+		y = stdlib_base_dists_rayleigh_kurtosis( sigma );
+		printf( "σ: %lf, kurtosis(X;σ): %lf\n", sigma, y );
```
```suggestion
		printf( "σ: %lf, Kurt(X,σ): %lf\n", sigma, y );
```

Planeshifter committed:
chore: apply suggestions from code review

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/kurtosis/benchmark/benchmark.js:
```diff
@@ -31,14 +32,20 @@ var kurtosis = require( './../lib' );
+	len = 100;
+	sigma = new Float64Array( len );
+	for ( i = 0; i < len; i++ ) {
+		sigma[ i ] = ( randu()*20.0 ) + EPS;
```
Should draw directly from `uniform` distribution.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/kurtosis/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,68 @@
+	len = 100;
+	sigma = new Float64Array( len );
+	for ( i = 0; i < len; i++ ) {
+		sigma[ i ] = ( randu()*20.0 ) + EPS;
```
Should draw directly from `uniform` distribution.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/kurtosis/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,139 @@
+	int i;
+
+	for ( i = 0; i < 100; i++ ) {
+		sigma[ i ] = random_uniform( 0.0, 20.0 ) + STDLIB_CONSTANT_FLOAT64_EPS;
```
```suggestion
		sigma[ i ] = random_uniform( STDLIB_CONSTANT_FLOAT64_EPS, 20.0 );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/kurtosis/src/main.c:
```diff
@@ -0,0 +1,62 @@
+#include "stdlib/math/base/assert/is_nan.h"
+#include "stdlib/constants/float64/pi.h"
+
+static const double KURTOSIS  = -( ( 6.0*STDLIB_CONSTANT_FLOAT64_PI*STDLIB_CONSTANT_FLOAT64_PI ) - ( 24.0*STDLIB_CONSTANT_FLOAT64_PI ) + 16.0 ) / ( (4.0-STDLIB_CONSTANT_FLOAT64_PI ) * ( 4.0-STDLIB_CONSTANT_FLOAT64_PI ) );
```
Can precompute this constant.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/kurtosis/src/main.c:
```diff
@@ -0,0 +1,62 @@
+/**
+* Returns the excess kurtosis of a Rayleigh distribution.
+*
+* @param {NonNegativeNumber} sigma - scale parameter
```
Should be C Doxygen instead of JSDoc.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/kurtosis/src/main.c:
```diff
@@ -0,0 +1,62 @@
+* @param {NonNegativeNumber} sigma - scale parameter
+* @returns {NonNegativeNumber} excess kurtosis
+*
+* @example
+* double v = stdlib_base_dists_rayleigh_kurtosis( 9.0 );
+* // returns ~0.245
```
Only keep this example.

Planeshifter committed:
chore: address PR feedback

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
chore: minor clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter reviewed (approved):
Let's get this one in.
